openmc/aggregate.py

```diff
@@ -0,0 +1,410 @@
+import sys
+from numbers import Integral
+
+import numpy as np
+
+from openmc import Filter, Nuclide
+from openmc.cross import CrossScore, CrossNuclide, CrossFilter
+from openmc.filter import _FILTER_TYPES
+import openmc.checkvalue as cv
+
+
+if sys.version_info[0] >= 3:
+    basestring = str
+
+# Acceptable tally aggregation operations
+_TALLY_AGGREGATE_OPS = ['sum', 'mean']
+
+
+class AggregateScore(object):
+    """A special-purpose tally score used to encapsulate an aggregate of a
+    subset or all of tally's scores for tally aggregation.
+
+    Parameters
+    ----------
+    scores : Iterable of str or CrossScore
+        The scores included in the aggregation
+    aggregate_op : str
+        The tally aggregation operator (e.g., 'sum', 'mean', etc.) used
+        to aggregate across a tally's scores with this AggregateScore
+
+    Attributes
+    ----------
+    scores : Iterable of str or CrossScore
+        The scores included in the aggregation
+    aggregate_op : str
+        The tally aggregation operator (e.g., 'sum', 'mean', etc.) used
+        to aggregate across a tally's scores with this AggregateScore
+
+    """
+
+    def __init__(self, scores=None, aggregate_op=None):
+
+        self._scores = None
+        self._aggregate_op = None
+
+        if scores is not None:
+            self.scores = scores
+        if aggregate_op is not None:
+            self.aggregate_op = aggregate_op
+
+    def __hash__(self):
+        return hash(repr(self))
+
+    def __eq__(self, other):
+        return str(other) == str(self)
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __deepcopy__(self, memo):
+        existing = memo.get(id(self))
+
+        # If this is the first time we have tried to copy this object, create a copy
+        if existing is None:
+            clone = type(self).__new__(type(self))
+            clone._scores = self.scores
+            clone._aggregate_op = self.aggregate_op
+
+            memo[id(self)] = clone
+
+            return clone
+
+        # If this object has been copied before, return the first copy made
+        else:
+            return existing
+
+    def __repr__(self):
+        string = ', '.join(map(str, self.scores))
```

Comment:
For some reason, when you sum across nuclides, it looks like this results in an extra comma (see Out[9] in the notebook). Doesn't make sense to me...